Automated response actions for rules

nastasha-solomon opened the pull request:
Fixes https://github.com/elastic/security-docs/issues/3488.

Preview:
- [Endpoint response action | Response actions history](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/response-actions.html#actions-log): Refreshed screenshot of an endpoint's response actions history log.
- [Response action history](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/response-actions-history.html): Refreshed screenshot and doc'd the new **Type** filter.
- [Host isolation | Isolate a host](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/host-isolation-ov.html#isolate-a-host): Added a fourth expandable section titled "Automatically isolate a host through a rule’s endpoint response action".
- [Create a detection rule | Set up response actions (optional)](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/rules-ui-create.html#rule-response-action): Added a blurb about the new Endpoint response action and refreshed screenshot.
- [View detection alerts](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/view-alert-details.html#view-alert-details): Updated the name and description of the fourth tab in the flyout (was formerly named the **Osquery Results** tab).
- [Add Osquery response action](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/osquery-response-action.html): Refreshed screenshots and updated multiple references to the former **Osquery Results** tab).
- [Create rule API | response actions schema](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/rules-api-create.html#response-actions-object-schema): Added docs for the Osquery response action and Endpoint response action.

tomsonpl reviewed (approved):
Looks great, good job, thanks 👍

caitlinbetz commented:
@nastasha-solomon only thought is that on the [Host isolation | Isolate a host](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/host-isolation-ov.html#isolate-a-host) page it may be beneficial to include a note about being thoughtful/intentional when configuring these rules. (Pasted the language previously discussed, essentially including something like the below) - 

> Host isolation involves quarantining a host from the network to prevent further spread of threats and limit potential damage. Please be advised that automatic host isolation based on a detection rule execution could result in unintended consequences, such as disrupting legitimate user activities or blocking critical business processes. 

otherwise looks great, thank you!

nastasha-solomon commented:
Thanks for the review and feedback, @caitlinbetz! I added warnings to the following sections:

- [Host isolation | Isolate a host](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/host-isolation-ov.html#isolate-a-host): Added warning to the fourth expandable section titled "Automatically isolate a host through a rule’s endpoint response action".
- [Create a detection rule | Set up response actions (optional)](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/rules-ui-create.html#rule-response-action): Added warning after the blurb about the new Endpoint response action.
- [Create rule API | response actions schema](https://security-docs_3546.docs-preview.app.elstc.co/guide/en/security/master/rules-api-create.html#response-actions-object-schema): Added a warning in the description for the `action_type_id` field. 

Let me know if this is what you were looking for and if you think the warning needs to be applied in additional places.

benironside commented on docs/detections/alerts-view-details.asciidoc:
```diff
@@ -1,6 +1,11 @@
 [[view-alert-details]]
 == View detection alert details
 
+:frontmatter-description: Create detection rules to monitor your environment for suspcious and malicious behavior.
```
```suggestion
:frontmatter-description: Create detection rules to monitor your environment for suspicious and malicious behavior.
```

benironside commented on docs/detections/alerts-view-details.asciidoc:
```diff
@@ -12,7 +17,7 @@ The alert details flyout contains these informational tabs:
 * *Table*: The alert data in table format. Data is organized into field-value pairs.
 * *JSON*: The alert data in JSON format.
-* *Osquery Results*: Results from queries attached to rules display on the *Osquery Results* tab. <<add-osquery-response-action,Learn more.>>
+* *Response Results*: Individually logged results from response actions. <<rule-response-action, Lean more.>>
```
```suggestion
* *Response Results*: Individually logged results from response actions. <<rule-response-action, Learn more.>>
```

benironside commented on docs/detections/api/rules/rules-api-create.asciidoc:
```diff
@@ -583,6 +588,50 @@ The following variables can only be passed if the rule’s action frequency is f
+* `query` (string, optional): To run a single query, use the `query` field and enter a  SQL query. Example: `"query": "SELECT * FROM processes;"` 
+* `saved_query_id` (string, optional): To run a saved query, use the `saved_query_id` field and specify the saved query ID. Example: `"saved_query_id":  "processes_elastic"`
+* `packId` (string, optional): To specify a query pack, use the `packId` field. Example: `"packId": "processes_elastic"`
+* `ecs_mapping` (object, required): Map Osquery results columns or static values to Elastic Common Schema (ECS) fields. Example: `"ecs_mapping": {"process.pid": {"field": "pid"}`
```
I noticed there are two open brackets { but only one close bracket. Just want to double-check this is right

nastasha-solomon replied:
Yep! You're right, I need to close the second bracket -- only one field is being mapped. Thanks!

```suggestion
* `ecs_mapping` (object, required): Map Osquery results columns or static values to Elastic Common Schema (ECS) fields. Example: `"ecs_mapping": {"process.pid": {"field": "pid"}}`
```

benironside commented on docs/detections/api/rules/rules-api-create.asciidoc:
```diff
@@ -583,6 +588,50 @@ The following variables can only be passed if the rule’s action frequency is f
+
+*Endpoint Security* 
+
+For Endpoint Security (`.endpoint`), specify an endpoint response action command and provide an optional a comment:
```
```suggestion
For Endpoint Security (`.endpoint`), specify an endpoint response action command and provide an optional comment:
```

benironside commented on docs/detections/rules-ui-create.asciidoc:
```diff
@@ -2,6 +2,11 @@
 [role="xpack"]
 == Create a detection rule
 
+:frontmatter-description: Create detection rules to monitor your environment for suspcious and malicious behavior.
```
```suggestion
:frontmatter-description: Create detection rules to monitor your environment for suspicious and malicious behavior.
```

benironside commented on docs/detections/rules-ui-create.asciidoc:
```diff
@@ -611,10 +616,12 @@ Example using the mustache "current element" notation `{{.}}` to output all the
 
+The endpoint response action allows you to automatically run response actions on an endpoint when rule conditions are met. Refer to <<isolate-a-host>> to learn more.
+
+IMPORTANT: Host isolation involves quarantining a host from the network to prevent further spread of threats and limit potential damage. Be aware that automatic host isolation from a detection rule execution could result in unintended consequences, such as disrupting legitimate user activities or blocking critical business processes.
```
```suggestion
IMPORTANT: Host isolation involves quarantining a host from the network to prevent further spread of threats and limit potential damage. Be aware that automatic host isolation can result in unintended consequences, such as disrupting legitimate user activities or blocking critical business processes.
```
Seemed like potential extra words, feel free to ignore

benironside replied:
if you like it, also applies in the collapsible block in host-isolation-ov.asciidoc

benironside commented on docs/management/admin/host-isolation-ov.asciidoc:
```diff
@@ -2,6 +2,11 @@
 [chapter, role="xpack"]
 = Host isolation
 
+:frontmatter-description: Host isolation allows you to isolate hosts from your network, blocking communication with other hosts on your network until you release the host. 
```
```suggestion
:frontmatter-description: Host isolation allows you to cut off a host's network access until you release it.
```

benironside commented on docs/management/admin/host-isolation-ov.asciidoc:
```diff
@@ -69,6 +74,21 @@ NOTE: The response console is an https://www.elastic.co/pricing[Enterprise subsc
 . Press *Return*.
 ====
 
+.Automatically isolate a host through a rule's endpoint response action
```
```suggestion
.Automatically isolate a host using a rule's endpoint response action
```

nastasha-solomon replied:
I went back and forth on this, so I'm glad you left this suggestion :)

benironside commented on docs/osquery/osquery-response-action.asciidoc:
```diff
@@ -2,6 +2,11 @@
 == Add Osquery Response Actions
 preview::[]
 
+:frontmatter-description: Osquery Response Actions allow you to add live queries to custom query rules so you can automatically collect data on systems the rule is monitoring. 
```
```suggestion
:frontmatter-description: Osquery Response Actions allow you to add live queries to custom query rules so you can automatically collect data on systems the rules are monitoring. 
```

benironside commented on docs/osquery/osquery-response-action.asciidoc:
```diff
@@ -59,7 +64,7 @@ IMPORTANT: If you edited a saved query or query pack that an Osquery Response Ac
 === Find query results
 
-When an alert is generated, Osquery automatically collects data on the system related to the alert. Query results are displayed within the *Osquery Results* tab in the Alert details flyout. The number next to the *Osquery Results* tab represents the number of queries attached to the rule.
+When an alert is generated, Osquery automatically collects data on the system related to the alert. Query results are displayed within the *Response Results* tab in the Alert details flyout. The number next to the *Response Results* tab represents the number of queries attached to the rule, in addition to endpoint response actions ran by the rule.
```
```suggestion
When a host generates an alert, Osquery automatically collects data about that host. Query results are displayed within the *Response Results* tab in the Alert details flyout. The number next to the *Response Results* tab represents the number of queries attached to the rule, in addition to endpoint response actions run by the rule.
```
Could be wrong here, but if the first sentence is accurate like this I think it's more clear (active voice too). Seems like "run" not "ran" since its not strictly about rules that ran in the past?

joepeeples replied:
I also like @benironside's active voice for the first sentence, but it's not quite accurate to say that a host generates an alert — a rule generates an alert, using data from the host. Maybe:
```suggestion
When a rule generates an alert, Osquery automatically collects data about the host involved. Query results are displayed within the *Response Results* tab in the Alert details flyout. The number next to the *Response Results* tab represents the number of queries attached to the rule, in addition to endpoint response actions run by the rule.
```
(I'm not 100% sold on "the host involved," could also be "the associated host" or "the affected host" or similar?)

And I'd go with `run`, because `ran` is just plain past tense but this sentence needs the past participle `run`, as in "actions _that were_ run by the rule" or "actions _that had been_ run by the rule."

nastasha-solomon replied:
I think we can make it more direct and just say that the Osquery action collects data on the host:

```suggestion
When an alert is generated, Osquery automatically collects data on the host. Query results are displayed within the *Response Results* tab in the Alert details flyout. The number next to the *Response Results* tab represents the number of queries attached to the rule, in addition to endpoint response actions ran by the rule.
```

benironside reviewed (commented):
Great work, big chunk of improvements here. Left some small suggestions, hope you like 'em.

nastasha-solomon commented on docs/management/admin/host-isolation-ov.asciidoc:
```diff
@@ -69,6 +74,21 @@ NOTE: The response console is an https://www.elastic.co/pricing[Enterprise subsc
+====
+NOTE: The host isolation endpoint response action is an https://www.elastic.co/pricing[Enterprise subscription] feature. 
+
+IMPORTANT: Automatic host isolation from a detection rule execution could result in unintended consequences, such as disrupting legitimate user activities or blocking critical business processes. Be aware of this as you're configuring rules. 
```
```suggestion
IMPORTANT: Be aware that automatic host isolation can result in unintended consequences, such as disrupting legitimate user activities or blocking critical business processes.
```

joepeeples reviewed (commented):
This PR is hot hot fire 🔥!! Massive effort across a lot of interrelated pages, I only had a few nits to add. Thanks so much for taking this one on!

nastasha-solomon commented on docs/detections/alerts-view-details.asciidoc:
```diff
@@ -12,7 +17,7 @@ The alert details flyout contains these informational tabs:
 * *Table*: The alert data in table format. Data is organized into field-value pairs.
 * *JSON*: The alert data in JSON format.
-* *Osquery Results*: Results from queries attached to rules display on the *Osquery Results* tab. <<add-osquery-response-action,Learn more.>>
+* *Response Results*: Individually logged results from response actions. <<rule-response-action, Learn more.>>
```
This edit works! I was mainly trying to convey that Osquery query results and Endpoint Security response action results were logged separately on the tab. Describing the format in which they're logged is ultimately less important than describing the tab's purpose though.
